File: 345.reverse-vowels-of-a-string.js
/*
 * @lc app=leetcode id=345 lang=javascript
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
  const res = s.split("");
  const vowels = "aeiouAEIOU";
  let i = 0;
  let j = s.length - 1;
  while (i < j) {
    if (!vowels.includes(res[i])) {
      ++i;
    } else if (!vowels.includes(res[j])) {
      --j;
    } else {
      [res[i], res[j]] = [res[j], res[i]];
      ++i, --j;
    }
  }
  return res.join("");
};
// @lc code=end

File: 345.reverse-vowels-of-a-string.go
/*
 * @lc app=leetcode id=345 lang=golang
 *
 * [345] Reverse Vowels of a String
 */

package leetcode

// @lc code=start
func reverseVowels(s string) string {
	b := []byte(s)
	for i, j := 0, len(s)-1; i < j; {
		if !helper345(b[i]) {
			i++
		} else if !helper345(b[j]) {
			j--
		} else {
			b[i], b[j] = b[j], b[i]
			i, j = i+1, j-1
		}
	}
	return string(b)
}

func helper345(c byte) bool {
	if c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' {
		return true
	}
	if c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U' {
		return true
	}
	return false
}

// @lc code=end
